Add preprocessor pass on visual shader when showing generated code

100gold opened the pull request:
Fixes #82555 
Add preprocessor pass and check for errors in visual shader editor.

100gold committed:
Add preprocessor pass on visual shader when showing generated code

bitsawer commented:
Looks pretty good. I would also add the preprocessor error file name, which makes it easier to find the actual issue if you have complex or nested includes.

Also, looks like the error line number highlight is off by one, needs `- 1`. Updated section of the code would be this:


```cpp
if (err != OK) {
    ERR_FAIL_COND(err_positions.size() == 0);

    String file = err_positions.front()->get().file;
    int err_line = err_positions.front()->get().line;
    Color error_line_color = EDITOR_GET("text_editor/theme/highlighting/mark_color");
    preview_text->set_line_background_color(err_line - 1, error_line_color);
    error_panel->show();

    error_label->set_text("error(" + file + ":" + itos(err_line) + "): " + error_pp);
    shader_error = true;
    return;
}
```

100gold commented:
> Looks pretty good. I would also add the preprocessor error file name, which makes it easier to find the actual issue if you have complex or nested includes.
> 
> Also, looks like the error line number highlight is off by one, needs `- 1`. Updated section of the code would be this:
> 
> ...

Thanks! There was another issue with the error line number when there was an error in the included file. Fixed both issues.

bitsawer commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -5083,20 +5084,52 @@ void VisualShaderEditor::_update_preview() {
+		String err_text;
+		Vector<ShaderLanguage::FilePosition> include_positions = sl.get_include_positions();
+		if (include_positions.size() > 1) {
+			//error is in an include
```
```suggestion
			// Error is in an include.
```

bitsawer reviewed (changes requested):
Tested locally and works as expected. One comment should be updated to the new format (capitalized and ends with a dot), but other than that looks good.

As a note to myself and future maintainers, this code is now pretty similar to ShaderTextEditor version with some small differences, so it might be possible to merge them in the future. No need to do that in this PR, however.

AThousandShips commented:
Your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

clayjohn reviewed (approved):
Seems fine to me. I am not familiar with this code, but I trust the other reviewers

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
